File: MorySnip/Classes/Layers/Layer_Action.vb
Public Class Layer_Action
    Inherits Layer

    Public Sub New(FirstPoint As Point, Action As Actions, Zone As Zones)
        Me.Action = Action
        Me.Zone = Zone
        Me.Start(FirstPoint)
    End Sub
    Public Property Action() As Actions
    Public Property Zone() As Zones
    ReadOnly Pen1 As New Pen(Color.White, 1)
    ReadOnly Pen2 As New Pen(Color.Black, 1) With {.DashStyle = Drawing2D.DashStyle.Custom, .DashPattern = New Single() {3, 3}}
    ReadOnly BrushFill As New Drawing2D.HatchBrush(Drawing2D.HatchStyle.BackwardDiagonal, Color.DarkGray, Color.Transparent)

    Public Overrides Sub Paint(g As Graphics)
        Dim r As Rectangle = Me.Bounds

        Select Case Me.Zone
            Case Zones.NotSelected
                If r.Y > 0 Then g.FillRectangle(Me.BrushFill, New Rectangle(0, 0, 10000, r.Y))
                If r.X > 0 Then g.FillRectangle(Me.BrushFill, New Rectangle(0, r.Y, r.X, 10000))
                g.FillRectangle(Me.BrushFill, New Rectangle(r.X, r.Y + r.Height, 10000, 10000))
                g.FillRectangle(Me.BrushFill, New Rectangle(r.X + r.Width, r.Y, 10000, r.Height))

            Case Zones.Selected
                g.FillRectangle(Me.BrushFill, r)
        End Select

        g.DrawRectangle(Me.Pen1, r)
        g.DrawRectangle(Me.Pen2, r)
    End Sub

    Public Overrides Sub Render(g As Graphics)
    End Sub
End Class

File: MorySnip/Classes/Layers/LayerAction.cs
using System.Drawing;
using System.Drawing.Drawing2D;
using Morysoft.MorySnip.Classes;

namespace Morysoft.MorySnip.Classes.Layers;

public class LayerAction : Layer
{
    public LayerAction(Point FirstPoint, Actions Action, Zones Zone)
    {
        this.Action = Action;
        this.Zone = Zone;
        this.Start(FirstPoint);
    }

    public Actions Action { get; set; }

    public Zones Zone { get; set; }

    private readonly Pen Pen1 = new(Color.White, 1);
    private readonly Pen Pen2 = new(Color.Black, 1) { DashStyle = DashStyle.Custom, DashPattern = new float[] { 3, 3 } };
    private readonly HatchBrush BrushFill = new(HatchStyle.BackwardDiagonal, Color.DarkGray, Color.Transparent);

    public override void Paint(Graphics g)
    {
        var r = this.Bounds;

        switch (this.Zone)
        {
            case Zones.NotSelected:
                {
                    if (r.Y > 0)
                    {
                        g.FillRectangle(this.BrushFill, new Rectangle(0, 0, 10000, r.Y));
                    }

                    if (r.X > 0)
                    {
                        g.FillRectangle(this.BrushFill, new Rectangle(0, r.Y, r.X, 10000));
                    }

                    g.FillRectangle(this.BrushFill, new Rectangle(r.X, r.Y + r.Height, 10000, 10000));
                    g.FillRectangle(this.BrushFill, new Rectangle(r.X + r.Width, r.Y, 10000, r.Height));
                    break;
                }

            case Zones.Selected:
                {
                    g.FillRectangle(this.BrushFill, r);
                    break;
                }
        }

        g.DrawRectangle(this.Pen1, r);
        g.DrawRectangle(this.Pen2, r);
    }

    public override void Render(Graphics g)
    {
    }

    public override bool IsValid => this.FirstPoint.X != this.LastPoint.X && this.FirstPoint.Y != this.LastPoint.Y;
}
